File: frontend/utils/indicators.ts
type Pattern = "Bullish" | "Bearish" | "Sideway";
type Signal = "BUY" | "SELL" | "NONE";

export type CandleInput = {
  open: number;
  high: number;
  low: number;
  close: number;
  prevOpen?: number;
  prevHigh?: number;
  prevLow?: number;
  prevClose?: number;
};

export type CandlePattern = {
  name: string;
  bias: "bullish" | "bearish" | "neutral";
};

/**
 * Detect common candlestick patterns from OHLC data.
 * Returns the most significant pattern found, or null.
 */
export function detectCandle(c: CandleInput): CandlePattern | null {
  const { open, high, low, close } = c;
  const body = Math.abs(close - open);
  const range = high - low;
  if (range === 0) return null;

  const upperShadow = high - Math.max(open, close);
  const lowerShadow = Math.min(open, close) - low;
  const bodyRatio = body / range;
  const isBullish = close > open;
  const isBearish = close < open;

  const prev = c.prevClose != null && c.prevOpen != null && c.prevHigh != null && c.prevLow != null
    ? { open: c.prevOpen, high: c.prevHigh, low: c.prevLow, close: c.prevClose }
    : null;
  const prevBody = prev ? Math.abs(prev.close - prev.open) : 0;

  // ── Two-candle patterns (check first, need prev) ──
  if (prev) {
    // Bullish Engulfing
    if (prev.close < prev.open && isBullish && open <= prev.close && close >= prev.open && body > prevBody) {
      return { name: "Bullish Engulfing", bias: "bullish" };
    }
    // Bearish Engulfing
    if (prev.close > prev.open && isBearish && open >= prev.close && close <= prev.open && body > prevBody) {
      return { name: "Bearish Engulfing", bias: "bearish" };
    }
    // Morning Star (simplified): prev bearish, small body, current bullish closing above prev midpoint
    if (prev.close < prev.open && isBullish && bodyRatio < 0.3 === false && close > (prev.open + prev.close) / 2 && body > prevBody * 0.5) {
      // skip — morning star needs 3 candles, handled via Hammer instead
    }
  }

  // ── Single-candle patterns ──

  // Doji: very small body
  if (bodyRatio < 0.05) {
    if (lowerShadow > upperShadow * 2 && lowerShadow > range * 0.3) {
      return { name: "Dragonfly Doji", bias: "bullish" };
    }
    if (upperShadow > lowerShadow * 2 && upperShadow > range * 0.3) {
      return { name: "Gravestone Doji", bias: "bearish" };
    }
    return { name: "Doji", bias: "neutral" };
  }

  // Hammer: small body at top, long lower shadow
  // Bullish context → Hammer; Bearish context → Hanging Man
  if (lowerShadow >= body * 2 && upperShadow < body * 0.5 && bodyRatio < 0.35) {
    return { name: isBullish ? "Hammer" : "Hanging Man", bias: isBullish ? "bullish" : "bearish" };
  }

  // Inverted Hammer / Shooting Star: small body at bottom, long upper shadow
  // Bullish context → Inverted Hammer; Bearish context → Shooting Star
  if (upperShadow >= body * 2 && lowerShadow < body * 0.5 && bodyRatio < 0.35) {
    return { name: isBullish ? "Inverted Hammer" : "Shooting Star", bias: isBullish ? "bullish" : "bearish" };
  }

  // Marubozu: almost no shadows
  if (upperShadow < range * 0.05 && lowerShadow < range * 0.05) {
    return { name: isBullish ? "Bullish Marubozu" : "Bearish Marubozu", bias: isBullish ? "bullish" : "bearish" };
  }

  // Spinning Top: small body, both shadows significant
  if (bodyRatio < 0.3 && upperShadow > body && lowerShadow > body) {
    return { name: "Spinning Top", bias: "neutral" };
  }

  return null;
}


export function sma(values: number[], window: number): number[] {
  if (window <= 1) {
    return [...values];
  }

  const out: number[] = [];
  let sum = 0;

  for (let i = 0; i < values.length; i += 1) {
    sum += values[i];
    if (i >= window) {
      sum -= values[i - window];
    }

    if (i < window - 1) {
      out.push(values[i]);
    } else {
      out.push(sum / window);
    }
  }

  return out;
}

/**
 * Exponential Moving Average.
 */
export function ema(values: number[], window: number): number[] {
  if (values.length === 0 || window <= 0) return [];
  const k = 2 / (window + 1);
  const out: number[] = [values[0]];
  for (let i = 1; i < values.length; i++) {
    out.push(values[i] * k + out[i - 1] * (1 - k));
  }
  return out;
}


// ── HalfTrend Indicator ──────────────────────────────────────────

export type HalfTrendPoint = {
  value: number;
  trend: 0 | 1; // 0 = uptrend (buy), 1 = downtrend (sell)
};

/**
 * Compute the HalfTrend indicator from OHLC data.
 * Pine Script port: amplitude controls highest/lowest lookback,
 * channelDeviation controls ATR-based channel width (price gap to half).
 */
export function halfTrend(
  bars: { high: number; low: number; close: number }[],
  amplitude: number = 2,
  channelDeviation: number = 10,
): (HalfTrendPoint | null)[] {
  const len = bars.length;
  if (len < 2) return new Array(len).fill(null);

  // ── ATR(100) via RMA ──
  const atrPeriod = 100;
  const tr: number[] = [bars[0].high - bars[0].low];
  for (let i = 1; i < len; i++) {
    const hl = bars[i].high - bars[i].low;
    const hc = Math.abs(bars[i].high - bars[i - 1].close);
    const lc = Math.abs(bars[i].low - bars[i - 1].close);
    tr.push(Math.max(hl, hc, lc));
  }
  const atr: number[] = [tr[0]];
  const alpha = 1 / atrPeriod;
  for (let i = 1; i < len; i++) {
    atr.push(alpha * tr[i] + (1 - alpha) * atr[i - 1]);
  }

  // ── SMA of highs and lows over amplitude ──
  function smaArr(vals: number[], period: number): number[] {
    const out: number[] = [];
    let sum = 0;
    for (let i = 0; i < vals.length; i++) {
      sum += vals[i];
      if (i >= period) sum -= vals[i - period];
      out.push(i < period - 1 ? vals[i] : sum / period);
    }
    return out;
  }
  const highma = smaArr(bars.map((b) => b.high), amplitude);
  const lowma = smaArr(bars.map((b) => b.low), amplitude);

  // ── Highest high / Lowest low over amplitude window ──
  function highest(vals: number[], period: number, idx: number): number {
    let mx = -Infinity;
    for (let j = Math.max(0, idx - period + 1); j <= idx; j++) mx = Math.max(mx, vals[j]);
    return mx;
  }
  function lowest(vals: number[], period: number, idx: number): number {
    let mn = Infinity;
    for (let j = Math.max(0, idx - period + 1); j <= idx; j++) mn = Math.min(mn, vals[j]);
    return mn;
  }

  const highs = bars.map((b) => b.high);
  const lows = bars.map((b) => b.low);

  // State variables (matching Pine var declarations)
  let trend = 0;         // 0 = up, 1 = down
  let nextTrend = 0;
  let maxLowPrice = lows[0];
  let minHighPrice = highs[0];
  let up = 0;
  let down = 0;

  const result: (HalfTrendPoint | null)[] = [];

  for (let i = 0; i < len; i++) {
    const atr2 = atr[i] / 2;
    const dev = channelDeviation * atr2;
    const highPrice = highs[highest_bar(highs, amplitude, i)];
    const lowPrice = lows[lowest_bar(lows, amplitude, i)];

    const prevTrend = trend;

    // HalfTrend Logic - Upward Trend Detection (nextTrend == 1)
    if (nextTrend === 1) {
      maxLowPrice = Math.max(lowPrice, maxLowPrice);
      if (highma[i] < maxLowPrice && bars[i].close < (i > 0 ? lows[i - 1] : lows[i])) {
        trend = 1;
        nextTrend = 0;
        minHighPrice = highPrice;
      }
    } else {
      // HalfTrend Logic - Downward Trend Detection (nextTrend == 0)
      minHighPrice = Math.min(highPrice, minHighPrice);
      if (lowma[i] > minHighPrice && bars[i].close > (i > 0 ? highs[i - 1] : highs[i])) {
        trend = 0;
        nextTrend = 1;
        maxLowPrice = lowPrice;
      }
    }

    // Calculate HalfTrend value (matching Pine exactly)
    if (trend === 0) {
      // Uptrend
      if (prevTrend !== 0) {
        // Trend just flipped to up: up = previous down value
        up = down;
      } else {
        up = i === 0 ? maxLowPrice : Math.max(maxLowPrice, up);
      }
      result.push({ value: up, trend: 0 });
    } else {
      // Downtrend
      if (prevTrend !== 1) {
        // Trend just flipped to down: down = previous up value
        down = up;
      } else {
        down = i === 0 ? minHighPrice : Math.min(minHighPrice, down);
      }
      result.push({ value: down, trend: 1 });
    }
  }

  return result;
}

/** Return index of highest value in window ending at idx */
function highest_bar(vals: number[], period: number, idx: number): number {
  let mxIdx = idx;
  for (let j = Math.max(0, idx - period + 1); j <= idx; j++) {
    if (vals[j] > vals[mxIdx]) mxIdx = j;
  }
  return mxIdx;
}

/** Return index of lowest value in window ending at idx */
function lowest_bar(vals: number[], period: number, idx: number): number {
  let mnIdx = idx;
  for (let j = Math.max(0, idx - period + 1); j <= idx; j++) {
    if (vals[j] < vals[mnIdx]) mnIdx = j;
  }
  return mnIdx;
}

type OHLCBar = { time: string; open: number; high: number; low: number; close: number };

export type SupertrendResult = {
  dir: 1 | -1;     // Pine convention: -1 = uptrend, 1 = downtrend
  value: number;    // supertrend line value
  flipUp: boolean;  // dir flipped from 1 → -1 (downtrend → uptrend)
  flipDown: boolean; // dir flipped from -1 → 1 (uptrend → downtrend)
};

/**
 * Aggregate daily OHLC into weekly bars (Mon–Fri).
 */
function aggregateWeekly(bars: OHLCBar[]): { weekly: OHLCBar[]; weekIndex: number[] } {
  const weekly: OHLCBar[] = [];
  const weekIndex: number[] = [];   // maps each daily bar to its weekly bar index

  let cur: OHLCBar | null = null;
  let wIdx = -1;

  for (let i = 0; i < bars.length; i++) {
    const d = new Date(bars[i].time);
    const dayOfWeek = d.getDay(); // 0=Sun .. 6=Sat

    // Start new week on Monday or if first bar
    const prevDate = i > 0 ? new Date(bars[i - 1].time) : null;
    const isNewWeek = cur === null || dayOfWeek === 1 ||
      (prevDate && d.getTime() - prevDate.getTime() > 3 * 86400000);

    if (isNewWeek) {
      if (cur) weekly.push(cur);
      cur = { time: bars[i].time, open: bars[i].open, high: bars[i].high, low: bars[i].low, close: bars[i].close };
      wIdx++;
    } else {
      cur!.high = Math.max(cur!.high, bars[i].high);
      cur!.low = Math.min(cur!.low, bars[i].low);
      cur!.close = bars[i].close;
    }
    weekIndex.push(wIdx);
  }
  if (cur) weekly.push(cur);

  return { weekly, weekIndex };
}

/**
 * Compute Supertrend on OHLC bars.
 * Pine Script logic: ATR period=10, multiplier=3.0
 */
function computeSupertrend(bars: OHLCBar[], period = 10, multiplier = 3.0): SupertrendResult[] {
  const len = bars.length;
  const results: SupertrendResult[] = [];

  // True Range
  const tr: number[] = [];
  for (let i = 0; i < len; i++) {
    if (i === 0) {
      tr.push(bars[i].high - bars[i].low);
    } else {
      const hl = bars[i].high - bars[i].low;
      const hc = Math.abs(bars[i].high - bars[i - 1].close);
      const lc = Math.abs(bars[i].low - bars[i - 1].close);
      tr.push(Math.max(hl, hc, lc));
    }
  }

  // ATR via RMA (Pine ta.rma / Wilder smoothing)
  // Pine: alpha = 1/length; sum := alpha*src + (1-alpha)*nz(sum[1])
  // Bar 0: nz(prev) = 0 → atr = alpha * tr
  const atr: number[] = new Array(len).fill(0);
  const alpha = 1.0 / period;
  for (let i = 0; i < len; i++) {
    const prev = i > 0 ? atr[i - 1] : 0;
    atr[i] = alpha * tr[i] + (1 - alpha) * prev;
  }

  // Supertrend (ta.supertrend convention: dir = -1 uptrend, dir = 1 downtrend)
  const up: number[] = new Array(len).fill(0);
  const dn: number[] = new Array(len).fill(0);
  const dir: (1 | -1)[] = new Array(len).fill(-1);

  for (let i = 0; i < len; i++) {
    const src = (bars[i].high + bars[i].low) / 2; // hl2
    const basicUp = src - multiplier * atr[i];
    const basicDn = src + multiplier * atr[i];

    if (i === 0) {
      up[i] = basicUp;
      dn[i] = basicDn;
      dir[i] = -1; // start uptrend
    } else {
      up[i] = bars[i - 1].close > up[i - 1] ? Math.max(basicUp, up[i - 1]) : basicUp;
      dn[i] = bars[i - 1].close < dn[i - 1] ? Math.min(basicDn, dn[i - 1]) : basicDn;

      // Pine: dir[1] == 1 (was downtrend) and close > dn[1] → flip to uptrend (-1)
      if (dir[i - 1] === 1 && bars[i].close > dn[i - 1]) {
        dir[i] = -1;
      } else if (dir[i - 1] === -1 && bars[i].close < up[i - 1]) {
        dir[i] = 1;
      } else {
        dir[i] = dir[i - 1];
      }
    }

    const flipUp = i > 0 && dir[i - 1] === 1 && dir[i] === -1;
    const flipDown = i > 0 && dir[i - 1] === -1 && dir[i] === 1;

    results.push({
      dir: dir[i],
      value: dir[i] === -1 ? up[i] : dn[i],
      flipUp,
      flipDown,
    });
  }

  return results;
}

/**
 * Compute weekly Supertrend and map back to daily bars.
 * Returns one SupertrendResult per daily bar.
 */
export function weeklySupertrend(
  dailyBars: OHLCBar[],
  period = 10,
  multiplier = 3.0,
): SupertrendResult[] {
  if (dailyBars.length === 0) return [];

  const { weekly, weekIndex } = aggregateWeekly(dailyBars);
  const wst = computeSupertrend(weekly, period, multiplier);

  // Map each daily bar to its weekly supertrend result
  return dailyBars.map((_, i) => wst[weekIndex[i]]);
}


export function detectPattern(price: number, sma20: number): Pattern {
  if (price > sma20) {
    return "Bullish";
  }
  if (price < sma20) {
    return "Bearish";
  }
  return "Sideway";
}


export function detectSignals(sma5: number[], sma10: number[]): Signal[] {
  const out: Signal[] = [];

  for (let i = 0; i < sma5.length; i += 1) {
    if (i === 0) {
      out.push("NONE");
      continue;
    }

    const prev5 = sma5[i - 1];
    const prev10 = sma10[i - 1];
    const curr5 = sma5[i];
    const curr10 = sma10[i];

    if (prev5 <= prev10 && curr5 > curr10) {
      out.push("BUY");
    } else if (prev5 >= prev10 && curr5 < curr10) {
      out.push("SELL");
    } else {
      out.push("NONE");
    }
  }

  return out;
}


/**
 * Detect abnormal/boost volume days.
 * A day is "boost" if volume >= threshold × rolling average volume.
 * Returns ratio (volume / avgVolume) for each day.
 */
export type VolumeInfo = {
  volume: number;
  avgVolume: number;
  ratio: number;
  boost: boolean;
};

export function detectVolumeBoost(
  volumes: number[],
  lookback: number = 20,
  threshold: number = 2.0,
): VolumeInfo[] {
  const out: VolumeInfo[] = [];
  for (let i = 0; i < volumes.length; i++) {
    const start = Math.max(0, i - lookback);
    const window = volumes.slice(start, i);
    const avg = window.length > 0 ? window.reduce((a, b) => a + b, 0) / window.length : 0;
    const ratio = avg > 0 ? volumes[i] / avg : 0;
    out.push({
      volume: volumes[i],
      avgVolume: Math.round(avg),
      ratio: Math.round(ratio * 100) / 100,
      boost: ratio >= threshold,
    });
  }
  return out;
}

File: klse_strategy/indicators.py
"""
indicators.py — Technical indicators for HalfTrend + Weekly Supertrend strategy.

Implements exact Pine Script logic:
  - HalfTrend with amplitude, channelDeviation, ATR deviation
  - Supertrend (ATR-based, factor/period configurable)
  - Weekly Supertrend aggregation (no lookahead)
  - ATR (Wilder RMA)
"""
from __future__ import annotations

import numpy as np
import pandas as pd


# ─── ATR (Wilder RMA) ─────────────────────────────────────────────────
def atr(highs: np.ndarray, lows: np.ndarray, closes: np.ndarray,
        period: int = 14) -> np.ndarray:
    """Average True Range using Wilder's smoothing (RMA)."""
    n = len(closes)
    tr = np.zeros(n)
    tr[0] = highs[0] - lows[0]
    for i in range(1, n):
        tr[i] = max(highs[i] - lows[i],
                     abs(highs[i] - closes[i - 1]),
                     abs(lows[i] - closes[i - 1]))
    alpha = 1.0 / period
    out = np.zeros(n)
    out[0] = tr[0]
    for i in range(1, n):
        out[i] = alpha * tr[i] + (1 - alpha) * out[i - 1]
    return out


# ─── SMA ───────────────────────────────────────────────────────────────
def sma(values: np.ndarray, period: int) -> np.ndarray:
    """Simple Moving Average."""
    s = pd.Series(values)
    return s.rolling(period, min_periods=1).mean().to_numpy()


# ─── EMA ───────────────────────────────────────────────────────────────
def ema(values: np.ndarray, period: int) -> np.ndarray:
    """Exponential Moving Average."""
    s = pd.Series(values)
    return s.ewm(span=period, adjust=False).mean().to_numpy()


# ─── Supertrend ────────────────────────────────────────────────────────
def supertrend(highs: np.ndarray, lows: np.ndarray, closes: np.ndarray,
               atr_period: int = 10, multiplier: float = 3.0
               ) -> tuple[np.ndarray, np.ndarray]:
    """
    Supertrend matching Pine Script ta.supertrend().

    Returns
    -------
    direction : ndarray  (-1 = uptrend/bullish, +1 = downtrend/bearish)
    st_line   : ndarray  supertrend value
    """
    n = len(closes)
    atr_vals = atr(highs, lows, closes, atr_period)

    hl2 = (highs + lows) / 2.0
    up_band = hl2 - multiplier * atr_vals
    dn_band = hl2 + multiplier * atr_vals

    final_up = np.zeros(n)
    final_dn = np.zeros(n)
    direction = np.ones(n)   # start bearish
    st_line = np.zeros(n)

    final_up[0] = up_band[0]
    final_dn[0] = dn_band[0]

    for i in range(1, n):
        # Persist upper band
        final_up[i] = max(up_band[i], final_up[i - 1]) \
            if closes[i - 1] > final_up[i - 1] else up_band[i]
        # Persist lower band
        final_dn[i] = min(dn_band[i], final_dn[i - 1]) \
            if closes[i - 1] < final_dn[i - 1] else dn_band[i]

        # Direction flip
        if direction[i - 1] == 1 and closes[i] > final_dn[i - 1]:
            direction[i] = -1
        elif direction[i - 1] == -1 and closes[i] < final_up[i - 1]:
            direction[i] = 1
        else:
            direction[i] = direction[i - 1]

        st_line[i] = final_up[i] if direction[i] == -1 else final_dn[i]

    return direction, st_line


def weekly_supertrend(daily_df: pd.DataFrame,
                      atr_period: int = 10,
                      multiplier: float = 3.0) -> np.ndarray:
    """
    Compute Weekly Supertrend and map back to daily bars WITHOUT lookahead.

    Each daily bar sees only the COMPLETED weekly Supertrend (prev week).
    The current week's partial data updates the weekly candle but the
    Supertrend direction used is from the *last completed* weekly bar
    until the current week closes.

    Returns ndarray (len == daily bars): -1 = bullish, +1 = bearish.
    """
    dates = daily_df["date"].values
    highs = daily_df["high"].values.astype(float)
    lows = daily_df["low"].values.astype(float)
    closes = daily_df["close"].values.astype(float)
    n = len(dates)

    # Aggregate daily → weekly
    w_highs, w_lows, w_closes = [], [], []
    week_map = np.zeros(n, dtype=int)  # daily idx → week idx
    w_idx = -1

    for i in range(n):
        dt = pd.Timestamp(dates[i])
        new_week = (i == 0
                    or dt.isocalendar()[1] != pd.Timestamp(dates[i - 1]).isocalendar()[1]
                    or dt.year != pd.Timestamp(dates[i - 1]).year)
        if new_week:
            w_idx += 1
            w_highs.append(highs[i])
            w_lows.append(lows[i])
            w_closes.append(closes[i])
        else:
            w_highs[w_idx] = max(w_highs[w_idx], highs[i])
            w_lows[w_idx] = min(w_lows[w_idx], lows[i])
            w_closes[w_idx] = closes[i]
        week_map[i] = w_idx

    w_h = np.array(w_highs)
    w_l = np.array(w_lows)
    w_c = np.array(w_closes)

    w_dir, _ = supertrend(w_h, w_l, w_c, atr_period, multiplier)

    # Map weekly direction back to daily (use PREVIOUS completed week to avoid lookahead)
    # Within current week, use dir from previous completed week
    out = np.ones(n)
    for i in range(n):
        wi = week_map[i]
        # Use the completed previous week's direction
        if wi > 0:
            out[i] = w_dir[wi - 1]
        else:
            out[i] = w_dir[0]  # first week, no prior data
    return out


# ─── HalfTrend (exact Pine Script replica) ────────────────────────────
def halftrend(highs: np.ndarray, lows: np.ndarray, closes: np.ndarray,
              amplitude: int = 5, channel_deviation: int = 2
              ) -> tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
    """
    Exact Pine Script HalfTrend indicator.

    Pine Script reference:
      amplitude = input(5)
      channelDeviation = input(2)
      Uses ATR(100) for channel width, SMA of high/low for trend detection.

    Parameters
    ----------
    amplitude : int
        Lookback for highest/lowest and SMA of high/low.
    channel_deviation : int
        Multiplier for ATR-based channel deviation.

    Returns
    -------
    trend       : ndarray  0 = uptrend (bullish), 1 = downtrend (bearish)
    ht_line     : ndarray  HalfTrend line value
    buy_signal  : ndarray  1 = buy signal (trend flipped bullish), 0 = no signal
    sell_signal : ndarray  1 = sell signal (trend flipped bearish), 0 = no signal
    """
    n = len(closes)
    trend = np.zeros(n, dtype=int)
    ht_line = np.zeros(n)
    buy_signal = np.zeros(n, dtype=int)
    sell_signal = np.zeros(n, dtype=int)

    # ATR(100) for deviation channel — matches Pine default
    atr_vals = atr(highs, lows, closes, 100)
    dev = channel_deviation * atr_vals

    next_trend = 0
    max_low_price = lows[0]
    min_high_price = highs[0]
    up = 0.0
    down = 0.0
    atr_high = 0.0
    atr_low = 0.0

    for i in range(n):
        lo = max(0, i - amplitude)
        highest_val = np.max(highs[lo:i + 1])
        lowest_val = np.min(lows[lo:i + 1])
        # SMA of highs and lows over amplitude period
        high_ma = np.mean(highs[lo:i + 1])
        low_ma = np.mean(lows[lo:i + 1])

        prev_trend = trend[i - 1] if i > 0 else next_trend

        if next_trend == 1:
            # Currently bearish, check for flip to bullish
            max_low_price = max(lowest_val, max_low_price)
            if high_ma < max_low_price and closes[i] < (lows[max(0, i - 1)] if i > 0 else lows[0]):
                trend[i] = 0  # flip to bullish
                next_trend = 0
                min_high_price = highest_val
                # Set up line
                up = max(max_low_price, up if i > 0 else max_low_price)
            else:
                trend[i] = 1  # stay bearish
        else:
            # Currently bullish, check for flip to bearish
            min_high_price = min(highest_val, min_high_price)
            if low_ma > min_high_price and closes[i] > (highs[max(0, i - 1)] if i > 0 else highs[0]):
                trend[i] = 1  # flip to bearish
                next_trend = 1
                max_low_price = lowest_val
                # Set down line
                down = min(min_high_price, down if (i > 0 and down > 0) else min_high_price)
            else:
                trend[i] = 0  # stay bullish

        # Update HalfTrend line
        if trend[i] == 0:
            if prev_trend != 0:
                up = max(max_low_price, up if i > 0 else max_low_price)
            else:
                up = max(max_low_price, up)
            atr_high = up + dev[i]
            atr_low = up - dev[i]
            ht_line[i] = up
        else:
            if prev_trend != 1:
                down = min(min_high_price, down if (i > 0 and down > 0) else min_high_price)
            else:
                down = min(min_high_price, down if down > 0 else min_high_price)
            atr_high = down + dev[i]
            atr_low = down - dev[i]
            ht_line[i] = down

        # Detect signals: trend flip
        if i > 0:
            if trend[i] == 0 and trend[i - 1] == 1:
                buy_signal[i] = 1   # bearish → bullish
            elif trend[i] == 1 and trend[i - 1] == 0:
                sell_signal[i] = 1  # bullish → bearish

    return trend, ht_line, buy_signal, sell_signal


# ─── Swing Low detection ──────────────────────────────────────────────
def swing_low(lows: np.ndarray, lookback: int = 10) -> np.ndarray:
    """Return rolling swing low (confirmed minimum over ±lookback bars)."""
    n = len(lows)
    out = np.full(n, np.nan)
    last_swing = np.nan
    for i in range(lookback, n - lookback):
        window = lows[i - lookback:i + lookback + 1]
        if lows[i] == np.min(window):
            last_swing = lows[i]
        out[i + lookback] = last_swing
    for i in range(1, n):
        if np.isnan(out[i]):
            out[i] = out[i - 1] if i > 0 else np.nan
    return out
